src/native/libs/System.Globalization.Native/pal_icushim_static.c

```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+        goto error;
+    }
+
+    struct stat src_st;
+    int result = fstat(src_fd, &src_st);
+    if (result == -1)
+    {
+        log_shim_error("apple_mmap_icu_data: failed to fstat %s", path);
+        goto error;
+    }
 
+    char tmp_file[APPLE_TMPFILE_NAME_SIZE];
+    size_t src_size = src_st.st_size;
+    size_t icu_data_size = 0;
+    const char *last_period = strrchr(path, '.');
+    const char *icu_data_name = NULL;
+    if (last_period && strcmp(last_period, ".lzfse") == 0)
+    {
+        char *cache_file = tmp_file;
+
+        uint64_t pid = (uint64_t) getpid();
+        int written = snprintf(tmp_file, APPLE_TMPFILE_NAME_SIZE, "%" PRIu64 ".", pid);
+        if (written < 0)
+        {
+            log_shim_error("apple_mmap_icu_data: failed to generate tmpfile PID prefix");
+            goto error;
+        }
+        cache_file += written;
+        written = snprintf(cache_file, APPLE_TMPFILE_NAME_SIZE - written, "icudt-%" PRIu64 "-%" PRIu64 "-%" PRIu64 ".dat.decompressed",
+            (uint64_t) src_st.st_ino,
+            (uint64_t) src_st.st_size,
+            (uint64_t) src_st.st_mtimespec.tv_sec);
+        if (written < 0)
+        {
+            log_shim_error("apple_mmap_icu_data: failed to generate cache file name");
+            goto error;
+        }
+
+        char cache_dir[PATH_MAX + 1];
+        result = apple_cache_directory(cache_dir);
+        if (result == -1)
+        {
+            goto error;
+        }
+        dir_fd = open(cache_dir, O_DIRECTORY);
+        if (dir_fd == -1)
+        {
+            log_shim_error("apple_mmap_icu_data: failed to open directory %s", cache_dir);
+            goto error;
+        }
+
+        cache_fd = openat(dir_fd, cache_file, O_RDONLY);
+        if (cache_fd == -1)
+        {
+            cache_fd = openat(dir_fd, tmp_file, O_RDWR | O_CREAT, 0640);
+            if (cache_fd == -1)
+            {
+                log_shim_error("apple_mmap_icu_data: failed to open %s/%s for writing", cache_dir, tmp_file);
+                goto error;
+            }
+            const char *src_mem = mmap(NULL, src_size, PROT_READ, MAP_SHARED, src_fd, 0);
+            if (src_mem == MAP_FAILED)
+            {
+                log_shim_error("apple_mmap_icu_data: failed to map %s with size %zu", path, src_size);
+                goto error;
+            }
+            size_t decompressed_data_size = 0;
+            int result = apple_decompress_to_fd(cache_fd, &decompressed_data_size, src_mem, src_st.st_size);
+            munmap((void *) src_mem, src_st.st_size);
+            fsync(cache_fd);
+            if (!result)
+            {
+                goto error;
+            }
+            result = renameat(dir_fd, tmp_file, dir_fd, cache_file);
+            fsync(dir_fd);
+            if (result == -1)
+            {
+                log_shim_error("apple_mmap_icu_data: failed to rename %s to %s", tmp_file, cache_file);
+                goto error;
+            }
```

Comment:
It doesn't look like you're deleting old files here; maybe iterate over all other "icudt-*.dat.decompressed" files and delete them?